framework/finalizer.go

```diff
@@ -0,0 +1,99 @@
+package framework
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	finalizerName = "operatorkit.giantswarm.io"
+)
+
+type patchSpec struct {
+	Op    string   `json:"op"`
+	Path  string   `json:"path"`
+	Value []string `json:"value"`
+}
+
+func (f *Framework) addFinalizer(obj interface{}) (bool, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return false, err
+	}
+	if containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		return false, nil // resource already has the finalizer.
+	}
+
+	c, err := rest.RESTClientFor(&rest.Config{})
+	if err != nil {
+		return false, err
+	}
+	patch := fmt.Sprintf(`{"op":"add","value":%q,"path":"/metadata/finalizers/1"}`, finalizerName)
```

Comment:
Talked briefly with Pawel about this, the JSONPatch should add the finalizer at the first index of the finalizers. This should preserve all other finalizers that are possibly set.

Reply:
`/1` at the end is disturbing. Is it index?

Reply:
We need tests to verify this. I think we can at least write unit tests for this.

Reply:
The `/1` is supposed to add the finalizer at the first position, see http://jsonpatch.com/#add

I agree on testing this.